src/uct/tcp/tcp.h

```diff
@@ -27,18 +33,29 @@ typedef struct uct_tcp_am_hdr {
 
 
 /**
- * TCP endpoint
+ * TCP endpoint communication context
  */
-typedef struct uct_tcp_ep {
-    uct_base_ep_t                 super;
-    int                           fd;        /* Socket file descriptor */
-    uint32_t                      events;    /* Current notifications */
-    ucs_queue_head_t              pending_q; /* Pending operations */
+typedef struct uct_tcp_ep_ctx {
     void                          *buf;      /* Partial send/recv data */
     size_t                        length;    /* How much data in the buffer */
     size_t                        offset;    /* Next offset to send/recv */
+    uct_tcp_ep_progress_t         progress;  /* Progress engine */
```

Comment:
why do we need progress pointers on iface?
is it just to detect errors?
could just add sume assertion before calling tx_progress / rx_progress, no need to bloat ep struct..

Reply:
This is needed to implement CM functions that will be utilized by an application thread:
- to progress non-blocking connection establishment, send a connection request to a peer
- to wait for a connection request after accepting a connection
- etc